tools/cache_creator/cache_info.cpp

```diff
@@ -0,0 +1,270 @@
+  if (auto err = privateHeaderOffsetOrErr.takeError())
+    return std::move(err);
+
+  return *privateHeaderOffsetOrErr + sizeof(vk::PipelineBinaryCachePrivateHeader);
+}
+
+// =====================================================================================================================
+// Reads all PipelineBinaryCache entries.
+//
+// @param [out] entriesInfoOut : The cache entries found
+// @param sourceElfMD5ToIdentifier : Optional map from source ELF MD5 sums to their identifiers.
+//                                   This allows to match cache entries to ELF files on disk, regardles of the hashing
+//                                   algorithm used for cache index keys.
+// @returns : Error if the cache blob does not have a valid content section, success otherwise
+llvm::Error CacheBlobInfo::readBinaryCacheEntriesInfo(llvm::SmallVectorImpl<BinaryCacheEntryInfo> &entriesInfoOut,
+                                                      llvm::StringMap<std::string> *sourceElfMD5ToIdentifier) const {
+  auto contentOffsetOrErr = getCacheContentOffset();
+  if (auto err = contentOffsetOrErr.takeError())
+    return err;
+
+  constexpr size_t EntrySize = sizeof(vk::BinaryCacheEntry);
+  const uint8_t *const blobStart = reinterpret_cast<const uint8_t *>(m_cacheBlob.getBufferStart());
+  const uint8_t *const blobEnd = reinterpret_cast<const uint8_t *>(m_cacheBlob.getBufferEnd());
+  const uint8_t *currData = reinterpret_cast<const uint8_t *>(m_cacheBlob.getBufferStart()) + *contentOffsetOrErr;
+  size_t entryIdx = 0;
+
+  while (currData < blobEnd) {
+    const size_t currEntryOffset = currData - blobStart;
+    if (currData + EntrySize > blobEnd)
+      return createBlobError(m_cacheBlob, "Insufficient buffer size for cache entry header #%zu at offset %zu",
+                             entryIdx, currEntryOffset);
+
+    BinaryCacheEntryInfo currEntryInfo = {};
+    currEntryInfo.entryHeader = reinterpret_cast<const vk::BinaryCacheEntry *>(currData);
+    currEntryInfo.idx = entryIdx;
+
+    const size_t currEntryBlobSize = currEntryInfo.entryHeader->dataSize;
+    if (currData + EntrySize + currEntryBlobSize > blobEnd)
+      return createBlobError(m_cacheBlob, "Insufficient buffer size for cache entry content #%zu at offset %zu",
+                             entryIdx, currEntryOffset);
+
+    currData += EntrySize;
+    currEntryInfo.entryBlob = {currData, currEntryBlobSize};
+    currData += currEntryBlobSize;
+
+    llvm::MD5 md5;
+    md5.update(currEntryInfo.entryBlob);
+    llvm::MD5::MD5Result result = {};
+    md5.final(result);
+    currEntryInfo.entryMD5Sum = result.digest();
+
+    if (sourceElfMD5ToIdentifier) {
+      auto sourceElfIt = sourceElfMD5ToIdentifier->find(currEntryInfo.entryMD5Sum);
+      if (sourceElfIt != sourceElfMD5ToIdentifier->end())
+        currEntryInfo.sourceElfBlobIdentifier = sourceElfIt->second;
+    }
+
+    currEntryInfo.sourceElfBlobIdentifier = llvm::None;
```

Comment:
Won't this overwrite the value stored on line 193?

Reply:
Yes, good catch.